be/src/storage/vectorized/row_source_mask.cpp

```diff
@@ -0,0 +1,147 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "storage/vectorized/row_source_mask.h"
+
+#include "common/config.h"
+#include "common/logging.h"
+#include "common/status.h"
+
+namespace starrocks::vectorized {
+
+RowSourceMaskBuffer::RowSourceMaskBuffer(int64_t tablet_id, const std::string& storage_root_path)
+        : _mask_column(std::move(UInt16Column::create_mutable())),
+          _tablet_id(tablet_id),
+          _storage_root_path(storage_root_path) {}
+
+RowSourceMaskBuffer::~RowSourceMaskBuffer() {
+    _reset_mask_column();
+    if (_tmp_file_fd > 0) {
+        ::close(_tmp_file_fd);
+    }
+}
+
+Status RowSourceMaskBuffer::write(const std::vector<RowSourceMask>& source_masks) {
+    if (_mask_column->byte_size() >= config::vertical_compaction_max_mask_memory_bytes) {
+        if (_tmp_file_fd == -1) {
+            RETURN_IF_ERROR(_create_tmp_file());
+        }
+        RETURN_IF_ERROR(_serialize_masks());
+        _reset_mask_column();
+    }
```

Comment:
Can we move this block after appending _maks_column? otherwise, the memory usage cannot be strained under `config::vertical_compaction_max_mask_memory_bytes`.

Reply:
done